### PR TITLE
[data] feat: decouple rmpad and dyn_bsz

### PR DESCRIPTION
**Pull request by yangtian6781**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `perf`, `ops`, `parallel`
  - If this PR involves multiple modules, separate them with `,` like `[ci, data, model]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

File: veomni/arguments/arguments_types.py [hunk 1]
```diff
@@ -704,7 +704,7 @@ def compute_train_steps(
         """
         Computes the training steps per epoch according to the data length.
         """
-        if self.rmpad or self.rmpad_with_pos_ids:
+        if (self.rmpad or self.rmpad_with_pos_ids) and self.dyn_bsz:
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/arguments/arguments_types.py]: @yangtian6781

**Comment by Luosuu** [comment]
@yangtian6781 Hi, thank you very much and please fix the tests. I think it is better to keep default behavior unchanged

**Comment by yangtian6781** [comment]
> @yangtian6781 Hi, thank you very much and please fix the tests. I think it is better to keep default behavior unchanged

Hi, thanks for your comment. I discard changes in tests.

**Comment by Luosuu** [comment]
Thank you very much @yangtian6781 
maybe there was some misunderstanding but I do would like to see new tests for 
* remove padding with position ids *with* dynamic batch sizes
* remove padding with position ids *without* dynamic batch sizes
(we only care about the case of rmpad_with_position_ids: True)
I think now we only test rmpad_with_pos_ids with dyn_bsz (implicit default).
Could you please also add a test for rmpad_with_pos_ids with dyn_bsz disabled?

File: veomni/arguments/arguments_types.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/arguments/arguments_types.py]: as I commented in PR thread, please add test cases for rmpad_with_pos_ids as True without dyn_bsz

**Comment by Luosuu** [comment]
Also it would be great to have some doc about this as well. Here is an example by codex:

Assume the dataset has 4 samples:

  - S1 = [A B C D] (len 4)
  - S2 = [E F] (len 2)
  - S3 = [G H I] (len 3)
  - S4 = [J] (len 1)

  ### rmpad_with_pos_ids = True, dyn_bsz = ON

  Goal: pack multiple samples into one long sequence and choose how many samples to fit per batch based on token budget.

  Example packed batch (one micro‑batch):

  - input_ids = [A B C D E F G H I J]
  - position_ids = [0 1 2 3 0 1 0 1 2 0]
  - (No padding added.)

  dyn_bsz decides how many samples to pack so total tokens ~ target budget. If your target is, say, 10 tokens, it packs all 4 here.

  ### rmpad_with_pos_ids = True, dyn_bsz = OFF

  Goal: still pack with position_ids, but batch size is fixed in number of samples, not by tokens.

  Example fixed batch size = 2 samples:

  - Batch 1: S1 + S2
    input_ids = [A B C D E F]
    position_ids = [0 1 2 3 0 1]
  - Batch 2: S3 + S4
    input_ids = [G H I J]
    position_ids = [0 1 2 0]

  So the difference is:

  - dyn_bsz ON: batch size varies to hit a token budget.
  - dyn_bsz OFF: batch size is fixed (#samples), but still uses position_ids packing (no padding).

**Comment by yangtian6781** [comment]
@Luosuu Thank you so much for your kindly guidance. I have added the corresponding tests and document.

File: tests/data/test_rmpad_with_pos_ids_without_dyn_bsz.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        "--train.global_batch_size=16",
+        "--train.micro_batch_size=2",
+        "--train.data_parallel_mode=ddp",
+        "--train.ckpt_manager=dcp",
+        f"--data.datasets_type={dataset_type}",
+        "--train.ulysses_parallel_size=1",
+        "--train.bsz_warmup_ratio=0",
+        "--train.output_dir=.tests/cache",
+        f"--train.rmpad=false",
+        f"--train.rmpad_with_pos_ids=true",
```
Comment (Luosuu) [inline tests/data/test_rmpad_with_pos_ids_without_dyn_bsz.py]: please fix the ruff linter. You can fix it by running pre-commit

File: docs/usage/rmpad_with_pos_ids_and_dyn_bsz.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Assume the dataset has 4 samples:
```
Comment (Luosuu) [inline docs/usage/rmpad_with_pos_ids_and_dyn_bsz.md]: Please give a proper title like other docs and enrich the doc a bit, like why we do rmpad with pos_ids (improve FA efficiency), why we have dyn_bsz config (pretrain vs. sft), ...

File: tests/data/test_rmpad_with_pos_ids_without_dyn_bsz.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
+import os
```
Comment (Luosuu) [inline tests/data/test_rmpad_with_pos_ids_without_dyn_bsz.py:13]: please explain your test design in comments a bit

Comment (yangtian6781) [reply tests/data/test_rmpad_with_pos_ids_without_dyn_bsz.py:13]: I copied this code: https://github.com/ByteDance-Seed/VeOmni/blob/main/tests/data/test_native_datasets.py

Then the following modifications were made:
1. Passed `args.train.dyn_bsz` to `build_dataloader`
2. Set `train.ulysses_parallel_size=1`, `train.rmpad=false`, and `train.rmpad_with_pos_ids=true` in the command
3. Added the `train.dyn_bsz` argument to the command to control `dyn_bsz`
4. Added the function `test_rmpad_with_pos_ids_without_dyn_bsz` to support pytest

It can be observed that `cu_seq_lens_q.shape` is `micro_batch_size+1` for each rank, as expected.

Comment (Luosuu) [reply tests/data/test_rmpad_with_pos_ids_without_dyn_bsz.py:13]: cool. yeah please add these explanation at the top of this test file so that others can know what this test does

**Comment by Luosuu** [comment]
thank you. let's wait for CI to finish and then I can approve & merge.

**Review by Luosuu — APPROVED** [review]
